### DISCUSSION TITLE
Cannot convert Document into HtmlDocument from an IFRAME content

### DISCUSSION THREAD

**Discussion by Yvo68** [wasm-bindgen/wasm-bindgen#3718] [Q&A]
### Describe the Bug

Conversion of Document into HTMLDocument wit dyn_into always fails for a IFRAME content document.
(works correctly on the main document)

### Steps to Reproduce

1. Have an iframe present in the DOM with id="edit" attribute
2. Run following code

let window = window().unwrap();
let document = window.document().unwrap();       
let elem = document.get_element_by_id("edit").unwrap();
let texteditor = elem.dyn_into::<HtmlIFrameElement>().unwrap();
let win2 = texteditor.content_window().unwrap();
let doc2 = win2.document();
let res_doc2_as_html = doc2.dyn_into::<HtmlDocument>();
match res_doc2_as_html {
                            Ok (doc2_as_html) => {
                                doc2_as_html.exec_command(pressed_id.as_str()).unwrap();
                            },
                            Err(_) => {
                                log!("Could not convert document into HTMLDocument");                            
                            }
                        }

Always returns the error (could not convert).

### Expected Behavior

The document should be converted into HtmlDocument so that command_exec can be called on the content.

### Actual Behavior

The dyn_into conversion fails.

### Additional Context

See demo Yew app attached.

[demo_err.zip](https://github.com/rustwasm/wasm-bindgen/files/13459456/demo_err.zip)

**Comment by daxpedda** [comment]
I didn't look into this, but I'm going to assume this is not an issue with `wasm-bindgen` but a API usage one.

I'm happy to look into it further if a minimal example is provided. With minimal I mean without any further dependencies, e.g. Yew.

**Comment by Yvo68** [comment]
Thanks for your answer.

Here is a small example without YEW.

The left IFRAME is accessed through RUST/WEBASM. The right IFRAME is accessed in JS (showing that it works).

[test_cmd.zip](https://github.com/rustwasm/wasm-bindgen/files/13467993/test_cmd.zip)

**Reply by daxpedda** [reply]
Please upload to a repository next time, it's not ideal to share files over GitHub unless necessary.

The problem is that `dyn_into()` uses `x instanceof HTMLDocument`, which fails because the `HTMLDocument` of the root window is different then the `IFrame`s one. So it would require `x instanceof iframeWindow.HTMLDocument`.

There is no good solution `wasm-bindgen`/`web-sys` can apply here, what you can do on your end is just use `unchecked_into()`.

**Reply by Yvo68** [reply]
Thanks.
The unchecked_into() conversion allows to make the send_command but the send_command has no effect on the IFRAME and returns false, while the send_command in the JS code works and returns true.

If the HTML documents of the root window and the document from the IFRAME are different, wouldn't you need to define different rust types (with shared traits). Wouldn't that be a good solution ?

**Reply by daxpedda** [reply]
> The unchecked_into() conversion allows to make the send_command but the send_command has no effect on the IFRAME and returns false, while the send_command in the JS code works and returns true.

Whats `send_command()`? Do you mean `exec_command()`?
I did try this locally and it worked just fine. `exec_command()` only works if text is actually selected, so this might be the issue.
> If the HTML documents of the root window and the document from the IFRAME are different, wouldn't you need to define different rust types (with shared traits). Wouldn't that be a good solution ?

The problem is that every `IFrame` has it's own types, this can't be represented in Rust because types have to be known at compile time and can't hold any values. A completely different API needs to be added to support this.

**Reply by Yvo68** [reply]
You're right, it works when a selection is present.
Thanks for your support.
